src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs

```diff
@@ -0,0 +1,265 @@
+            lease = null;
+            return false;
+        }
+
+        private void Release(int releaseCount)
+        {
+            lock (Lock)
+            {
+                _permitCount += releaseCount;
+                Debug.Assert(_permitCount <=  _options.PermitLimit);
+
+                while (_queue.Count > 0)
+                {
+                    RequestRegistration nextPendingRequest =
+                        _options.QueueProcessingOrder == QueueProcessingOrder.OldestFirst
+                        ? _queue.PeekHead()
+                        : _queue.PeekTail();
+
+                    if (_permitCount >= nextPendingRequest.Count)
+                    {
+                        nextPendingRequest =
+                            _options.QueueProcessingOrder == QueueProcessingOrder.OldestFirst
+                            ? _queue.DequeueHead()
+                            : _queue.DequeueTail();
+
+                        _permitCount -= nextPendingRequest.Count;
+                        _queueCount -= nextPendingRequest.Count;
+                        Debug.Assert(_queueCount >= 0);
+                        Debug.Assert(_permitCount >= 0);
+
+                        ConcurrencyLease lease = nextPendingRequest.Count == 0 ? SuccessfulLease : new ConcurrencyLease(true, this, nextPendingRequest.Count);
+                        // Check if request was canceled
+                        if (!nextPendingRequest.Tcs.TrySetResult(lease))
+                        {
+                            // Queued item was canceled so add count back
+                            _permitCount += nextPendingRequest.Count;
+                        }
+                        nextPendingRequest.CancellationTokenRegistration.Dispose();
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+            }
+        }
+
+        private class ConcurrencyLease : RateLimitLease
+        {
+            private bool _disposed;
+            private readonly ConcurrencyLimiter? _limiter;
+            private readonly int _count;
+            private readonly string? _reason;
+
+            public ConcurrencyLease(bool isAcquired, ConcurrencyLimiter? limiter, int count, string? reason = null)
+            {
+                IsAcquired = isAcquired;
+                _limiter = limiter;
+                _count = count;
+                _reason = reason;
+
+                // No need to set the limiter if count is 0, Dispose will noop
+                Debug.Assert(count == 0 ? limiter is null : true);
+            }
+
+            public override bool IsAcquired { get; }
+
+            public override IEnumerable<string> MetadataNames => Enumerable();
+
+            private IEnumerable<string> Enumerable()
+            {
+                if (_reason is not null)
+                {
+                    yield return MetadataName.ReasonPhrase.Name;
+                }
+            }
```

Comment:
```suggestion
            public override IEnumerable<string> MetadataNames => _reason is null ?
                Array.Empty<string>() :
                new[] { MetadataName.ReasonPhrase.Name };
```
?  I'm not sure how MetadataNames is used.

Reply:
I'm not 100% sure we're using it as intended currently. I'm thinking it's intended to be a static list of metadata names that this lease _can_ have, not what it currently has.